src/protocols/internal/encode.c

```diff
@@ -244,32 +221,24 @@ static ssize_t internal_encode(fr_dbuff_t *dbuff,
 	/*
 	 *	Encode the total length, and write the width
 	 *	of the integer to the encoding byte.
-	 *
-	 *	Already did length checks at the start of
-	 *	the function.
 	 */
-	{
-		value_end = dbuff->p;
-		flen = fr_net_from_uint64v(buff, value_end - value_field);
+	slen = fr_dbuff_marker_current(&value_end) - fr_dbuff_marker_current(&value_field);
+	flen = fr_net_from_uint64v(buff, slen);
 
-		/*
-		 *	Ugh, it's a long one, need to memmove the
-		 *	data, good job we lied to all the encoding
-		 *      functions earlier, so we *must* have enough
-		 *	buffer space.
-		 */
-		if (flen > 1) {
-			fr_dbuff_advance(dbuff, flen - 1);
-			memmove(len_field + flen, value_field, value_end - value_field);
-			value_end += flen - 1;
-		}
-		memcpy(len_field, buff, flen);
-		enc_field[0] |= ((flen - 1) << 2);
+	/*
+	 *	Ugh, it's a long one, need to memmove the
+	 *	data.
+	 */
+	if (flen > 1) {
+		FR_DBUFF_ADVANCE_RETURN(&work_dbuff, flen - 1);
+		memmove(fr_dbuff_marker_current(&value_field) + flen - 1, fr_dbuff_marker_current(&value_field), slen);
```

Comment:
This then leaves the current pointer in work_dbuff set incorrectly, which isn't a problem now, but could well be if we ever tried to refactor.  You need to add a dbuff function that operates on a marker or sbuff, and updates all pointers within the segment being moved. The function would also need to check that no markers or `p` are moved past the end of the buffer.  If you don't feel confident in writing such a function then I can take a look.